### PR TITLE
8093 error callback

### LINKED ISSUE #8093: Error Callback
The error function callback provided by EnergyPlus https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/DataGlobals.hh#L296 does not inform the severity of the error. Consider adding an argument to indicate error level.

### PR DESCRIPTION
**Pull request by kbenne**
Fixes #8093

### CODE REVIEW

**Comment by nrel-bot-3** [comment]
@kbenne @lgentile it has been 28 days since this pull request was last updated.

**Comment by kbenne** [comment]
@Myoldmopar this is additive on top of https://github.com/NREL/EnergyPlus/pull/8086

File: src/EnergyPlus/DataGlobals.cc [hunk 1]
```diff
@@ -251,9 +251,9 @@ namespace DataGlobals {
     int Progress(0); // current progress (0-100)
     void (*fProgressPtr)(int const);
     void (*fMessagePtr)(std::string const &);
-    void (*progressCallback)(int const);
-    void (*messageCallback)(const char * message);
-    void (*errorCallback)(const char * errorMessage);
+    std::function<void(int const)> progressCallback;
+    std::function<void(const std::string &)> messageCallback;
+    std::function<void(EnergyPlus::Error e, const std::string &)> errorCallback;
```
Comment (Myoldmopar) [inline src/EnergyPlus/DataGlobals.cc:256]: I do like moving to using std::function as long as we are inside the C++ boundary.

File: src/EnergyPlus/api/runtime.cc [hunk 2]
```diff
@@ -104,16 +104,23 @@ void registerProgressCallback(EnergyPlusState EP_UNUSED(state), void (*f)(int co
...
+void registerStdOutCallback(EnergyPlusState EP_UNUSED(state), std::function<void (const std::string &)> f) {
     EnergyPlus::DataGlobals::messageCallback = f;
 }
 
+void registerStdOutCallback(EnergyPlusState state, void (*f)(const char * message)) {
+    const auto stdf = [f](const std::string & message) {
+        f(message.c_str());
+    };
+    registerStdOutCallback(state, std::function<void (const std::string &)>(stdf));
+}
```
Comment (Myoldmopar) [inline src/EnergyPlus/api/runtime.cc:96]: I feel like really we should invert all the register callbacks to this approach.

File: src/EnergyPlus/api/TypeDefs.h [hunk 3]
```diff
@@ -51,4 +51,18 @@
...
+enum class Error: int {
+  Continue = 0,
+  Info = 1,
+  Warning = 2,
+  Severe = 3,
+  Fatal = 4
+};
+
+} // namespace EnergyPlus
+#endif
```
Comment (Myoldmopar) [inline src/EnergyPlus/api/TypeDefs.h:66]: So if you are accessing this endpoint in C, or Python, you'll just need to hardcode the numbers?  I do like the use of a proper enum class inside C++, but what are your thoughts on balancing that nicety vs having to hardwire numbers elsewhere.  If we went with a plain C enum, then we could access it nicely through the API.

**Review by Myoldmopar — COMMENTED** [review]
I'd like to chat about the enum mostly, otherwise this is pretty good.

File: src/EnergyPlus/api/TypeDefs.h (hunk 3 above)
Comment (kbenne) [reply src/EnergyPlus/api/TypeDefs.h:66]: Yeah this certainly does prioritize C++, which to me is a feature not a bug but that is subjective. The C api is just an int https://github.com/NREL/EnergyPlus/blob/8093-error-callback/src/EnergyPlus/api/func.h#L94 

A C style enum would be ok, but what would the enum values be named? There are no namespaces so you would probably want something like ENERGYPLUS_ERROR_CONTINUE, or something like that. Perhaps with less uppercase shouting.

@lefticus you got any input?

**Comment by Myoldmopar** [comment]
OK, I got develop pulled in, but I think I might like to move the error callback into the runtime API section where all the other callback registration functions live.  Any issue with that?

**Comment by Myoldmopar** [comment]
OK, reproduced the errors locally and fixed them.  I anticipate this will be ready to go later today after CI verifies.

**Comment by Myoldmopar** [comment]
I'm not going to wait on integration debug to pass, this change is totally unrelated.  Here you go @kbenne, merging in.

**Pull request merged**